### PR TITLE
Customize concurrency for ESSource DataProxies

### LINKED ISSUE #36487: Concurrent fetching of condition data
CMSSW currently limits fetching of condition data (from the ESSource) to one thread at the time.

With improvements to add thread safery to the frontier_client as well as the Coral library it has been discussed that implementation allowing parallel execution of requests could reduce time spent in beginning-of-run transitions (in HLT and possibly offline) since these jobs are multithreaded.
An important point to also assess if the limiting the parallelism is necessary for each online HLT as well as offline system. Main issue with this is to avoid loading Squid servers, therefore this option can allow tuning the amount of load.

This ticket is opened to track the issue and more details will (could) be added to it.

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Refactored ESSource specific DataProxies to allow customizing how concurrency should be handled. Provided implementations to support either concurrent or non-concurrent running of DataProxies for the same ESSource.

This was ability will be used by PoolDBESSource.

#### PR validation:

Code compiles and all framework unit tests (including a new one) all succeed.

closes #36487

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by perrotta** [comment]
please test
(just to check whether in the new IB the differences in 11634.911 can disappear)

File: FWCore/Framework/interface/ESSourceDataProxyBase.h [hunk 1]
```diff
@@ -53,19 +45,46 @@ namespace edm::eventsetup {
 
     virtual void prefetch(edm::eventsetup::DataKey const& iKey, EventSetupRecordDetails) = 0;
 
+    //Should call from prefetchAsyncImpl
+    template <typename ASYNC, typename GUARD>
+    void prefetchAsyncImplTemplate(ASYNC iAsync,
+                                   GUARD iGuard,
```
Comment (makortel) [inline FWCore/Framework/interface/ESSourceDataProxyBase.h:51]: How about
```suggestion
                                   GUARD iGuardFactory,
```
for more consistent naming with `guardPrefetch()`?

File: FWCore/Framework/interface/ESSourceDataProxyBase.h [hunk 2]
```diff
@@ -53,19 +45,46 @@ namespace edm::eventsetup {
...
+                                   edm::WaitingTaskHolder iTask,
+                                   edm::eventsetup::EventSetupRecordImpl const& iRecord,
+                                   edm::eventsetup::DataKey const& iKey,
+                                   edm::ESParentContext const& iContext) {
+      auto group = iTask.group();
+      if (needToPrefetch(std::move(iTask))) {
+        iAsync(*group, [this, iGuard, &iRecord, iKey, iContext]() {
```
Comment (makortel) [inline FWCore/Framework/interface/ESSourceDataProxyBase.h:58]:
```suggestion
        iAsync(*group, [this, iGuardFactory, &iRecord, iKey, iContext]() {
```

File: FWCore/Framework/interface/ESSourceDataProxyBase.h [hunk 3]
```diff
@@ -53,19 +45,46 @@ namespace edm::eventsetup {
...
+          try {
+            guardPrefetch(iGuard, iRecord, iKey, iContext);
```
Comment (makortel) [inline FWCore/Framework/interface/ESSourceDataProxyBase.h:60]:
```suggestion
            guardPrefetch(iGuardFactory, iRecord, iKey, iContext);
```

File: FWCore/Framework/interface/ESSourceDataProxyBase.h (hunk 1 above)
Comment (Dr15Jones) [reply FWCore/Framework/interface/ESSourceDataProxyBase.h]: done

File: FWCore/Framework/interface/ESSourceDataProxyBase.h (hunk 3 above)
Comment (Dr15Jones) [reply FWCore/Framework/interface/ESSourceDataProxyBase.h]: done

File: FWCore/Framework/interface/ESSourceDataProxyBase.h (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Framework/interface/ESSourceDataProxyBase.h]: done

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
merge

**Comment by santocch** [comment]
+1

**Pull request merged**